File: test-suite/src/test/java/io/micronaut/docs/jdbc/multiple/InventoryService.java
package io.micronaut.docs.jdbc.multiple;

// tag::imports[]
import io.micronaut.transaction.annotation.Transactional;
import jakarta.inject.Inject;
import jakarta.inject.Named;
import jakarta.inject.Singleton;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;
// end::imports[]

// tag::clazz[]
@Singleton
public class InventoryService {

    @Inject
    DataSource dataSource; // <1>

    @Inject
    @Named("warehouse")
    DataSource warehouseDataSource; // <2>

    @Transactional // <3>
    public String defaultUrl() throws SQLException {
        return url(dataSource);
    }

    @Transactional("warehouse") // <4>
    public String warehouseUrl() throws SQLException {
        return url(warehouseDataSource);
    }

    private static String url(DataSource dataSource) throws SQLException {
        try (Connection connection = dataSource.getConnection()) {
            return connection.getMetaData().getURL();
        }
    }
}
// end::clazz[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/jdbc/multiple/InventoryService.kt
package io.micronaut.docs.jdbc.multiple

// tag::imports[]
import io.micronaut.transaction.annotation.Transactional
import jakarta.inject.Inject
import jakarta.inject.Named
import jakarta.inject.Singleton
import javax.sql.DataSource
// end::imports[]

// tag::clazz[]
@Singleton
open class InventoryService {

    @Inject
    lateinit var dataSource: DataSource // <1>

    @Inject
    @Named("warehouse")
    lateinit var warehouseDataSource: DataSource // <2>

    @Transactional // <3>
    open fun defaultUrl(): String = url(dataSource)

    @Transactional("warehouse") // <4>
    open fun warehouseUrl(): String = url(warehouseDataSource)

    private fun url(dataSource: DataSource): String =
        dataSource.connection.use { connection -> connection.metaData.url }
}
// end::clazz[]
